File: tests/benchmarks/coro/skynet/skynet.erl
%% Skynet benchmark - Erlang equivalent
%% Classic coroutine tree benchmark: each node spawns 10 children
%% Leaf nodes return their sequence number, intermediate nodes sum children
%% Usage: erl -noshell -s skynet main -s init stop
-module(skynet).
-export([main/0, main/1]).

main() ->
    main(6).

main(Depth) ->
    Size = trunc(math:pow(10, Depth)),
    TotalCoros = calc_total(Depth),
    io:format("=== Skynet 测试 (Erlang) ===~n"),
    io:format("深度: ~p~n", [Depth]),
    io:format("预计协程数: ~p~n", [TotalCoros]),

    T0 = erlang:monotonic_time(millisecond),
    Result = skynet(0, Size, 10),
    T1 = erlang:monotonic_time(millisecond),

    Elapsed = T1 - T0,
    Expected = Size * (Size - 1) div 2,
    io:format("结果: ~p~n", [Result]),
    io:format("预期: ~p~n", [Expected]),
    io:format("正确: ~p~n", [Result =:= Expected]),
    io:format("时间: ~p ms~n", [Elapsed]),
    io:format("协程数: ~p~n", [TotalCoros]).

skynet(Num, 1, _Div) ->
    Num;
skynet(Num, Size, Div) ->
    ChildSize = Size div Div,
    Parent = self(),
    [spawn(fun() ->
        R = skynet(Num + I * ChildSize, ChildSize, Div),
        Parent ! R
    end) || I <- lists:seq(0, Div - 1)],
    collect_sum(Div, 0).

collect_sum(0, Sum) ->
    Sum;
collect_sum(N, Sum) ->
    receive Val ->
        collect_sum(N - 1, Sum + Val)
    end.

calc_total(Depth) ->
    calc_total(Depth, 0, 1).
calc_total(-1, Total, _N) ->
    Total;
calc_total(D, Total, N) ->
    calc_total(D - 1, Total + N, N * 10).

File: tests/benchmarks/coro/skynet/skynet.go
// 作者：xingleixu@gmail.com
// Skynet 基准测试
// 经典的协程性能测试，创建树形结构的 goroutine
// 每个节点创建10个子节点，叶节点返回自己的序号
// 最终汇总所有叶节点的值

package main

import (
	"fmt"
	"os"
	"strconv"
	"time"
)

func skynet(c chan int64, num int64, size int64, div int64) {
	if size == 1 {
		c <- num
		return
	}

	rc := make(chan int64, div)
	childSize := size / div

	for i := int64(0); i < div; i++ {
		childNum := num + i*childSize
		go skynet(rc, childNum, childSize, div)
	}

	var sum int64
	for i := int64(0); i < div; i++ {
		sum += <-rc
	}

	c <- sum
}

func calcTotal(depth int) int64 {
	var total int64 = 0
	var n int64 = 1
	for i := 0; i <= depth; i++ {
		total += n
		n *= 10
	}
	return total
}

func main() {
	depth := 6 // 深度6 = 10^6 = 100万协程

	if len(os.Args) > 1 {
		if d, err := strconv.Atoi(os.Args[1]); err == nil {
			depth = d
		}
	}

	fmt.Println("=== Skynet 测试 ===")
	fmt.Println("深度:", depth)
	fmt.Println("预计协程数:", calcTotal(depth))

	var size int64 = 1
	for i := 0; i < depth; i++ {
		size *= 10
	}

	c := make(chan int64, 1)

	start := time.Now()

	go skynet(c, 0, size, 10)
	result := <-c

	elapsed := time.Since(start)
	elapsedMs := float64(elapsed) / float64(time.Millisecond)

	// 预期结果: 0 + 1 + 2 + ... + (size-1) = size * (size - 1) / 2
	expected := size * (size - 1) / 2

	fmt.Println("结果:", result)
	fmt.Println("预期:", expected)
	fmt.Println("正确:", result == expected)
	fmt.Printf("时间: %.3f ms\n", elapsedMs)
	fmt.Println("协程数:", calcTotal(depth))
}
